### PR TITLE
System wide remotes

### LINKED ISSUE #7322: Implement system-wide remotes
The CLI tool (`lxc`) reads its config from `~/.config/lxc/config.yml`, that's where the various remotes are defined except for the built-in ones.

Users on shared systems have shown interest in adding remotes to all users.
For this, we'd need the LXD CLI to read both its normal config as well as a global configuration.

The apply order would effectively be:
 - Built-in
 - System
 - User

With each level being able to override the other, resulting in a complete config.
When saving, only what's specific to the user should be saved. Users will be able to override system remotes at which point they'll just be added to their own config.

On top of having a global `config.yml`, we also will need a global `servercerts` directory to host potential certificates for the global remotes. If a remote is marked as coming from the global config, we'd look for certs in that path rather than the usual `~/.config/lxc/servercerts`.

### PR DESCRIPTION
**Pull request by presztak**
This resolves #7322

### CODE REVIEW

File: lxc/config/config.go [hunk 1]
```diff
@@ -36,6 +36,14 @@ type Config struct {
 	cookieJars map[string]*cookiejar.Jar
 }
 
+// GlobalConfigPath returns a joined path of the global configuration directory and passed arguments
+func (c *Config) GlobalConfigPath(paths ...string) string {
+	path := []string{"/etc/lxd/config"}
```
Comment (stgraber) [inline lxc/config/config.go:41]: Probably should be just `/etc/lxd`

File: lxc/config/file.go [hunk 2]
```diff
@@ -34,6 +34,22 @@ func LoadConfig(path string) (*Config, error) {
 		}
 	}
 
+	//Apply the global (system-wide) remotes
```
Comment (stgraber) [inline lxc/config/file.go:37]: Missing space after beginning of comment

File: lxc/config/remote.go [hunk 3]
```diff
@@ -27,6 +27,7 @@ type Remote struct {
 	Project  string `yaml:"project,omitempty"`
 	Protocol string `yaml:"protocol,omitempty"`
 	Public   bool   `yaml:"public"`
+	Global   bool   `yaml:"gloabl,omitempty"`
```
Comment (stgraber) [inline lxc/config/remote.go:30]: This shouldn't ever be stored in config so should be `yaml:"-"` to keep it as an internal flag like Static

File: lxc/remote.go [hunk 4]
```diff
@@ -631,9 +637,18 @@ func (c *cmdRemoteRename) Run(cmd *cobra.Command, args []string) error {
...
-		if err != nil {
-			return err
+		if conf.Remotes[args[0]].Global == true {
+			err := conf.CopyGlobalCert(args[0], args[1])
+			if err != nil {
+				return err
+			}
+			rc.Global = false
+		} else {
+
```
Comment (stgraber) [inline lxc/remote.go:647]: Un-needed blank line

File: doc/remotes.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# Remotes
+
+## Introduction
+
```
Comment (stgraber) [inline doc/remotes.md:4]: Un-needed blank line

File: doc/remotes.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
+# Remotes
+
```
Comment (stgraber) [inline doc/remotes.md:2]: Un-needed blank line

File: doc/remotes.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+## Introduction
+
+Remotes (machine with LXD installed) is concept used by LXD through which users doesn't have to always provide hostname or IP and then validating certificate information.
+LXD introduce three types of remotes:
+
```
Comment (stgraber) [inline doc/remotes.md:7]: Should this be:
 - Static (hardcoded)
 - Local (per-user)
 - Global (per-system)

Instead?

File: doc/remotes.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+- local (default)
+- images
+- ubuntu
+- ubuntu-daily
+
+Local, ubuntu, ubuntu-daily are static and they can't be modified by user.
+
+### System
+
+System remotes can be managed only manually. Remote configuration path is: `/etc/lxd/config/config.yml`.
```
Comment (stgraber) [inline doc/remotes.md:24]: Maybe word this as:

"System remotes can be managed manually by editing `/etc/lxd/config.yml`."?

File: doc/remotes.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+System remotes certifications should be placed in `/etc/lxd/config/servercerts/` and named after remote name e.g. `foo.crt`
```
Comment (stgraber) [inline doc/remotes.md:25]: Should be:
"System remote certificates should be placed in `/etc/lxd/servercerts/` and be named after their respective remote. e.g. `foo.crt`"

File: doc/remotes.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    addr: https://10.0.2.5:8443
+    auth_type: tls
+    project: default
+    protocol: lxd
+    public: false
+    
+```
+
+### User
+
```
Comment (stgraber) [inline doc/remotes.md:46]: Un-needed blank line

File: doc/remotes.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+User level remotes are managed from CLI (`lxc`) with command:
+`lxc remote [command]`
+
+By default configuration file is kept in `~/.config/lxc/config.yml` or in `LXD_CONF` if defined.
```
Comment (stgraber) [inline doc/remotes.md:50]: *the configuration file

File: doc/remotes.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    auth_type: tls
+    project: default
+    protocol: lxd
+    public: false
+    
+```
+
+### User
+
+User level remotes are managed from CLI (`lxc`) with command:
```
Comment (stgraber) [inline doc/remotes.md:47]: *from the CLI (`lxc`) with:

File: doc/remotes.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+`lxc remote [command]`
+
+By default configuration file is kept in `~/.config/lxc/config.yml` or in `LXD_CONF` if defined.
+User have possibility to override system remotes (e.g. by running command `lxc remote rename` or `lxc remote set-url`)
```
Comment (stgraber) [inline doc/remotes.md:51]: *Users have the possibility
*e.g. by running `lxc remote rename`

File: doc/remotes.md [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+which result to add configuration to their own config. Certification files are copied also.
```
Comment (stgraber) [inline doc/remotes.md:52]: "which results in the remote being copied to their own config, including any associated certificates."

**Comment by stgraber** [comment]
Looking good, just a few nits and grammar/wording updates for the doc.

Thanks!

File: lxc/config/cert.go [hunk 15]
```diff
@@ -1,7 +1,10 @@
 package config
 
 import (
+	"fmt"
 	"github.com/lxc/lxd/shared"
```
Comment (stgraber) [inline lxc/config/cert.go:5]: We usually separate the builtin imports and the external ones, so this should become:
```
fmt
io
os

github.com/lxc/lxd/shared
```

File: lxc/config/config.go [hunk 16]
```diff
@@ -36,6 +36,14 @@ type Config struct {
 	cookieJars map[string]*cookiejar.Jar
 }
 
+// GlobalConfigPath returns a joined path of the global configuration directory and passed arguments
+func (c *Config) GlobalConfigPath(paths ...string) string {
+	path := []string{"/etc/lxd"}
```
Comment (stgraber) [inline lxc/config/config.go:41]: Can you make this consider `LXD_GLOBAL_CONF` environment and use that over `/etc/lxd/` if set?

That should line up with what we do for `LXD_CONF` and will be needed for the snap package.

Comment (presztak) [reply lxc/config/config.go]: @stgraber is it ok? Or should we add new `GlobalConfigDir` field to [Config](https://github.com/lxc/lxd/blob/master/lxc/config/config.go#L11)?

**Comment by stgraber** [comment]
jenkins: test this please

File: doc/remotes.md [hunk 17]
```diff
@@ -0,0 +1,49 @@
+# Remotes
+## Introduction
+Remotes (machine with LXD installed) is concept used by LXD through which users doesn't have to always provide hostname or IP and then validating certificate information.
```
Comment (stgraber) [inline doc/remotes.md:3]: Maybe rephrase as:
```
Remotes are a concept in the LXD command line client which are used to refer to various LXD servers or clusters.
A remote is effectively a name pointing to the URL of a particular LXD server as well as needed credentials to login and authenticate the server.
```

File: doc/remotes.md [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+LXD introduce four types of remotes:
```
Comment (stgraber) [inline doc/remotes.md:4]: `LXD has four types of remotes:`

File: doc/remotes.md [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+- Global (per-system)
+- Local (per-user)
+
+### Static
+Static remotes are:
+- local (default)
+- ubuntu
+- ubuntu-daily
+
+They are hardcoded and can't be modified by user.
```
Comment (stgraber) [inline doc/remotes.md:17]: *by the user

File: doc/remotes.md [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+
+### Default
+Automatically added on first use.
+
+### Global (per-system)
+Global remotes can be managed manually by editing /etc/lxd/config.yml.
+Global remote certificates should be placed in /etc/lxd/servercerts/ and be named after their respective remote. e.g. `foo.crt`
```
Comment (stgraber) [inline doc/remotes.md:24]:
```
By default the global configuration file is kept in `/etc/lxc/config.yml` or in `LXD_GLOBAL_CONF` if defined.
The configuration file can be manually edited to add global remotes. Certificates for those remotes should be stored inside the `servercerts` directory (e.g. /etc/lxc/servercerts/) and match the remote name (e.g. `foo.crt`).
```

**Comment by stgraber** [comment]
Just a few minor tweaks to the doc and we'll be good to go!

**Comment by stgraber** [comment]
jenkins: test this please
